File: solacher/src/jobs/markets.js
require('dotenv').config()

const { parentPort: workerParent, workerData } = require('node:worker_threads')

const redis = require('@redis/client')
const request = require('got-cjs').got

function calculateChange(oldVal, newVal) {
  if (newVal === oldVal || oldVal === 0) return 0;
  else return -(((oldVal - newVal) / oldVal) * 100);
}

;(async function () {
  const redisUrl = workerParent
    ? workerData.data.redisUrl
    : process.env.REDIS_URL

  const [rawPriceData, rawHistoryData, rawTvlData] = await Promise.all([
    request('https://api.coingecko.com/api/v3/simple/price', {
      searchParams: {
        ids: 'solana',
        vs_currencies: 'usd',
        include_market_cap: true,
        include_24hr_vol: true,
        include_24hr_change: true,
      },
    }).json(),
    request('https://api.coingecko.com/api/v3/coins/solana/market_chart', {
      searchParams: {
        vs_currency: 'usd',
        days: 13,
        interval: 'daily',
      },
    }).json(),
    request('https://api.llama.fi/charts/Solana').json(),
  ])

  const normalizedMarketsData = {
    data: {
      price: parseFloat(rawPriceData.solana.usd.toFixed(2)),
      tvl: rawTvlData[rawTvlData.length - 1].totalLiquidityUSD,
      tvlChange: parseFloat(
        calculateChange(
          rawTvlData[rawTvlData.length - 2].totalLiquidityUSD,
          rawTvlData[rawTvlData.length - 1].totalLiquidityUSD,
        ).toFixed(2),
      ),
      volume: Math.round(rawPriceData.solana.usd_24h_vol),
      change: parseFloat(rawPriceData.solana.usd_24h_change.toFixed(1)),
      marketCap: Math.round(rawPriceData.solana.usd_market_cap),
      history: rawHistoryData.prices.map((_, n) => ({
        ts: rawHistoryData.prices[n][0],
        price: rawHistoryData.prices[n][1],
        volume: rawHistoryData.total_volumes[n][1],
      })),
      tvlHistory: rawTvlData.map((item) => ({
        ts: parseInt(item.date) * 1000,
        tvl: item.totalLiquidityUSD,
      })),
    },
    count: null,
    type: 'object',
    updatedAt: Date.now(),
  }

  const redisClient = redis.createClient({ url: redisUrl })
  await redisClient.connect()
  await redisClient.set('markets', JSON.stringify(normalizedMarketsData))
  await redisClient.quit()

  if (workerParent) workerParent.postMessage('done')
  else process.exit(0)
})();

File: etl/etl/jobs/markets.py
import json
import os

import httpx
from dotenv import load_dotenv
from toolz import merge

from etl.connector import get_redis_connection
from etl.utils import calc_change, now, s_to_ms, str_format_money


def normalize_price_history(raw_history_data):
    price_history = []
    reduced_raw_history = raw_history_data["prices"][0:-1]
    for n, raw_price in enumerate(reduced_raw_history):
        price_history.append(
            {
                "ts": raw_price[0],
                "price": round(raw_price[1], 2),
                "volume": round(raw_history_data["total_volumes"][n][1]),
            }
        )
    return price_history


def normalize_tvl_history(raw_tvl_data):
    tvl_history = []
    reduced_raw_tvl_history = raw_tvl_data[-14:]
    for raw_tvl in reduced_raw_tvl_history:
        tvl_history.append(
            {
                "ts": s_to_ms(raw_tvl["date"]),
                "tvl": round(raw_tvl["totalLiquidityUSD"]),
            }
        )
    return tvl_history


def merge_price_tvl_history(price_history, tvl_history):
    merged_history = []
    for n, price in enumerate(price_history):
        merged_history.append(merge(price, tvl_history[n]))
    return merged_history


def normalize_markets_data(raw_price_data, raw_price_history, raw_tvl_history):
    price_history = normalize_price_history(raw_price_history)
    tvl_history = normalize_tvl_history(raw_tvl_history)
    merged_history = merge_price_tvl_history(price_history, tvl_history)

    return {
        "price": str_format_money(raw_price_data["quote"]["USD"]["price"]),
        "change": round(raw_price_data["quote"]["USD"]["percent_change_24h"], 2),
        "tvl": raw_tvl_history[-1]["totalLiquidityUSD"],
        "tvlChange": round(
            calc_change(
                raw_tvl_history[-2]["totalLiquidityUSD"],
                raw_tvl_history[-1]["totalLiquidityUSD"],
            ),
            2,
        ),
        "volume": round(raw_price_data["quote"]["USD"]["volume_24h"]),
        "volumeChangePercent": round(raw_price_data["quote"]["USD"]["volume_change_24h"], 2),
        "marketCap": round(raw_price_data["quote"]["USD"]["market_cap"]),
        "marketCapRank": raw_price_data["cmc_rank"],
        "history": merged_history,
    }


def update_markets():
    redis = get_redis_connection()

    price_response = httpx.get(
        "https://pro-api.coinmarketcap.com/v2/cryptocurrency/quotes/latest",
        params={
            "id": 5426,
            "aux": "cmc_rank",
        },
        headers={
            "X-CMC_PRO_API_KEY": os.environ["CMC_TOKEN"],
        },
    )
    price_response.raise_for_status()
    raw_price_data = price_response.json()["data"]["5426"]

    history_response = httpx.get(
        "https://api.coingecko.com/api/v3/coins/solana/market_chart",
        params={
            "vs_currency": "usd",
            "days": 14,
            "interval": "daily",
        },
    )
    history_response.raise_for_status()
    raw_history_data = history_response.json()

    tvl_response = httpx.get("https://api.llama.fi/charts/Solana")
    tvl_response.raise_for_status()
    raw_tvl_data = tvl_response.json()

    result = json.dumps(
        {
            "data": normalize_markets_data(raw_price_data, raw_history_data, raw_tvl_data),
            "count": 0,
            "type": "object",
            "updatedAt": now(),
        }
    )

    redis.set("markets", result)
    redis.close()

    return result


if __name__ == "__main__":
    load_dotenv()
    print(update_markets())
